File: test/python-tests/test_tuples.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


int main() {
    std::cout << "=== Tuple Creation ===" << std::endl;
    auto empty_tuple = ();
    auto single_tuple = (1,);
    auto pair_tuple = (1, 2);
    auto multi_tuple = (1, 2, 3, 4, 5);
    auto mixed_tuple = (1, "hello", 3.14, true);
    std::cout << to_str("Empty tuple: ") << to_str(empty_tuple) << std::endl;
    std::cout << to_str("Single tuple: ") << to_str(single_tuple) << std::endl;
    std::cout << to_str("Pair tuple: ") << to_str(pair_tuple) << std::endl;
    std::cout << to_str("Multi tuple: ") << to_str(multi_tuple) << std::endl;
    std::cout << to_str("Mixed tuple: ") << to_str(mixed_tuple) << std::endl;
    auto tuple_no_parens = 1, 2, 3, 4;
    std::cout << to_str("Tuple without parens: ") << to_str(tuple_no_parens) << std::endl;
    std::cout << "\n=== Tuple Indexing and Slicing ===" << std::endl;
    auto numbers = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
    std::cout << to_str("First element: ") << to_str(numbers[0]) << std::endl;
    std::cout << to_str("Last element: ") << to_str(numbers[-1]) << std::endl;
    std::cout << to_str("Slice [2:5]: ") << to_str(numbers[2) << std::endl;
    std::cout << to_str("Slice [:3]: ") << to_str(numbers[) << std::endl;
    std::cout << to_str("Slice [5:]: ") << to_str(numbers[5) << std::endl;
    std::cout << to_str("Slice [::2]: ") << to_str(numbers[) << std::endl;
    std::cout << to_str("Slice [::-1]: ") << to_str(numbers[) << std::endl;
    std::cout << "\n=== Tuple Unpacking ===" << std::endl;
    auto point = (10, 20);
    auto x, y = point;
    std::cout << to_str("x = ") << to_str(x) << to_str(", y = ") << to_str(y) << std::endl;
    auto a, b, c = 1, 2, 3;
    std::cout << to_str("a = ") << to_str(a) << to_str(", b = ") << to_str(b) << to_str(", c = ") << to_str(c) << std::endl;
    auto a, b = b, a;
    std::cout << to_str("After swap: a = ") << to_str(a) << to_str(", b = ") << to_str(b) << std::endl;
    auto nested = ((1, 2), (3, 4));
    auto (a, b), (c, d) = nested;
    std::cout << to_str("Nested unpacking: a=") << to_str(a) << to_str(", b=") << to_str(b) << to_str(", c=") << to_str(c) << to_str(", d=") << to_str(d) << std::endl;
    std::cout << "\n=== Tuple Operations ===" << std::endl;
    auto tuple1 = (1, 2, 3);
    auto tuple2 = (4, 5, 6);
    auto combined = tuple1 + tuple2;
    std::cout << to_str("Concatenation: ") << to_str(combined) << std::endl;
    auto repeated = tuple1 * 3;
    std::cout << to_str("Repetition: ") << to_str(repeated) << std::endl;
    std::cout << to_str("Length of tuple1: ") << to_str(len(tuple1)) << std::endl;
    std::cout << to_str("2 in tuple1: ") << to_str(2 in tuple1) << std::endl;
    std::cout << to_str("10 in tuple1: ") << to_str(10 in tuple1) << std::endl;
    std::cout << "\n=== Tuple Methods ===" << std::endl;
    auto numbers = (1, 2, 3, 2, 4, 2, 5);
    std::cout << to_str("Count of 2: ") << to_str(numbers.count(2)) << std::endl;
    std::cout << to_str("Index of 3: ") << to_str(numbers.index(3)) << std::endl;
    std::cout << to_str("Index of first 2: ") << to_str(numbers.index(2)) << std::endl;
    std::cout << "\n=== Tuples as Dictionary Keys ===" << std::endl;
    auto locations = {;
    };
    std::cout << to_str("Location at (0, 0): ") << to_str(locations[(0, 0)]) << std::endl;
    std::cout << to_str("Location at (1, 0): ") << to_str(locations[(1, 0)]) << std::endl;
    std::cout << "\n=== Nested Tuples ===" << std::endl;
    auto matrix = (;
    );
    std::cout << to_str("Matrix: ") << to_str(matrix) << std::endl;
    std::cout << to_str("Element [1][2]: ") << to_str(matrix[1][2]) << std::endl;
    std::cout << "\n=== Tuple Comparison ===" << std::endl;
    auto t1 = (1, 2, 3);
    auto t2 = (1, 2, 3);
    auto t3 = (1, 2, 4);
    std::cout << to_str("t1 == t2: ") << to_str(t1 == t2) << std::endl;
    std::cout << to_str("t1 == t3: ") << to_str(t1 == t3) << std::endl;
    std::cout << to_str("t1 < t3: ") << to_str(t1 < t3) << std::endl;
    std::cout << "\n=== Type Conversion ===" << std::endl;
    auto my_list = std::vector<decltype(1)>{1, 2, 3, 4};
    auto my_tuple = tuple(my_list);
    std::cout << to_str("List to tuple: ") << to_str(my_tuple) << std::endl;
    auto back_to_list = list(my_tuple);
    std::cout << to_str("Tuple to list: ") << to_str(back_to_list) << std::endl;
    std::cout << "\n=== All tuple tests completed ===" << std::endl;
    return 0;
}

File: test/python-tests/test_tuples.py
# Test tuple operations

# Tuple creation
print("=== Tuple Creation ===")
empty_tuple = ()
single_tuple = (1,)  # Note the comma
pair_tuple = (1, 2)
multi_tuple = (1, 2, 3, 4, 5)
mixed_tuple = (1, "hello", 3.14, True)

print(f"Empty tuple: {empty_tuple}")
print(f"Single tuple: {single_tuple}")
print(f"Pair tuple: {pair_tuple}")
print(f"Multi tuple: {multi_tuple}")
print(f"Mixed tuple: {mixed_tuple}")

# Tuple without parentheses
tuple_no_parens = 1, 2, 3, 4
print(f"Tuple without parens: {tuple_no_parens}")

# Tuple indexing and slicing
print("\n=== Tuple Indexing and Slicing ===")
numbers = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
print(f"First element: {numbers[0]}")
print(f"Last element: {numbers[-1]}")
print(f"Slice [2:5]: {numbers[2:5]}")
print(f"Slice [:3]: {numbers[:3]}")
print(f"Slice [5:]: {numbers[5:]}")
print(f"Slice [::2]: {numbers[::2]}")
print(f"Slice [::-1]: {numbers[::-1]}")

# Tuple unpacking
print("\n=== Tuple Unpacking ===")
point = (10, 20)
x, y = point
print(f"x = {x}, y = {y}")

# Multiple assignment
a, b, c = 1, 2, 3
print(f"a = {a}, b = {b}, c = {c}")

# Swap values
a, b = b, a
print(f"After swap: a = {a}, b = {b}")

# Nested tuple unpacking
nested = ((1, 2), (3, 4))
(a, b), (c, d) = nested
print(f"Nested unpacking: a={a}, b={b}, c={c}, d={d}")

# Tuple operations
print("\n=== Tuple Operations ===")
tuple1 = (1, 2, 3)
tuple2 = (4, 5, 6)

# Concatenation
combined = tuple1 + tuple2
print(f"Concatenation: {combined}")

# Repetition
repeated = tuple1 * 3
print(f"Repetition: {repeated}")

# Length
print(f"Length of tuple1: {len(tuple1)}")

# Membership
print(f"2 in tuple1: {2 in tuple1}")
print(f"10 in tuple1: {10 in tuple1}")

# Tuple methods
print("\n=== Tuple Methods ===")
numbers = (1, 2, 3, 2, 4, 2, 5)
print(f"Count of 2: {numbers.count(2)}")
print(f"Index of 3: {numbers.index(3)}")
print(f"Index of first 2: {numbers.index(2)}")

# Tuple as dictionary key (immutable)
print("\n=== Tuples as Dictionary Keys ===")
locations = {
    (0, 0): "origin",
    (1, 0): "east",
    (0, 1): "north"
}
print(f"Location at (0, 0): {locations[(0, 0)]}")
print(f"Location at (1, 0): {locations[(1, 0)]}")

# Nested tuples
print("\n=== Nested Tuples ===")
matrix = (
    (1, 2, 3),
    (4, 5, 6),
    (7, 8, 9)
)
print(f"Matrix: {matrix}")
print(f"Element [1][2]: {matrix[1][2]}")

# Tuple comparison
print("\n=== Tuple Comparison ===")
t1 = (1, 2, 3)
t2 = (1, 2, 3)
t3 = (1, 2, 4)
print(f"t1 == t2: {t1 == t2}")
print(f"t1 == t3: {t1 == t3}")
print(f"t1 < t3: {t1 < t3}")

# Convert list to tuple and vice versa
print("\n=== Type Conversion ===")
my_list = [1, 2, 3, 4]
my_tuple = tuple(my_list)
print(f"List to tuple: {my_tuple}")

back_to_list = list(my_tuple)
print(f"Tuple to list: {back_to_list}")

print("\n=== All tuple tests completed ===")
